### DISCUSSION TITLE
RemoveAsync doens't remove item from the Database.

### DISCUSSION THREAD

**Discussion by orchardc** [OrchardCMS/OrchardCore#8133] [Q&A]
The following code doens't remove item from the Database.
`await _contentManager.RemoveAsync(contentItem);`

Physically remove contents from the database - is it even possible with API? Otherwise the database is growing very fast if not removable. 

Any suggestion?

**Comment by hishamco** [comment]
Orchard Core is using soft delete, so no the content is not physically deleted

**Comment by Skrypt** [comment]
You can clean those records manually by using YesSQL directly like this :  `_session.Delete(existingContentItem)`

**Reply by orchardc** [reply]
@Skrypt , do you meaning using c# code with _session.Delete ??

The following code doesn't do anything with the data (_session is from YesSql.ISession )

```
                    foreach (var contentItem in contentItems)
                    {
                        //await _contentManager.RemoveAsync(contentItem);
                        _session.Delete(contentItem);
                    }
```

**Reply by Skrypt** [reply]
Impossible. At least it should remove the data from the ContentItemIndex table. Unless your contentItem doesn't have a Id that is matching on in your database.

**Reply by orchardc** [reply]
I counted in SQL before and after running the above code, the result is the same.

Note this is a backgroundTask, will it make any difference? 

`select count(*) from ContentItemIndex
`

**Reply by Skrypt** [reply]
Well are you removing an actual ContentItem that has previously been persisted in the database or not? 
If you are using a background task that creates a Content Item and that you want to delete it right after it might not work because the INSERT query might never have happened yet. YesSql session methods are no op. They get executed on dispose of the current scope. You can't create/delete a content item in the same background task.

Edit : 

Please look for await _session.CommitAsync(); to force a session to be committed.

**Reply by orchardc** [reply]
Thanks @Skrypt , What I really try to do is to physically remove the records from the db. 

For example, I want to delete all content items that its
**ContentType = "MyProduct"**
 and 
**Category = "Trail"** (a field in MyProduct)

Ideally, it can be done programmtically via Background Task or Workflow.

Is it possible?

**Reply by Skrypt** [reply]
We can't filter by Fields unless you created a custom index on them. Or if you used the ContentFieldIndexing feature in OC.
Else, you will need to use JSON functions in SQL to filter. Or maybe a GraphQL Query would work "I never tried that approach yet".
But from the _session.Query<SomeIndex>(). Unless you previously indexed these fields ; you won't be able to filter them.

So, you need to at least get the proper filtered Identifiers to remove them.

So you have many options. The faster one being a SQL Query using JSON functions "I think".

https://docs.microsoft.com/en-us/sql/t-sql/functions/json-functions-transact-sql?view=sql-server-ver15

**Reply by orchardc** [reply]
Can we filter by ContentType only? If so, how to programmtically (via Background Task or Workflow) delete the following records?

`ContentType = "MyProduct"`

Many thanks!!

**Reply by Skrypt** [reply]
`var contentItems = await _session.Query<ContentItem, ContentItemIndex>(x => x.ContentType == "MyProduct")`

**Reply by orchardc** [reply]
Thanks @Skrypt , but there is an error with the code:
![image](https://user-images.githubusercontent.com/16193516/103042201-bd543c00-4570-11eb-9d15-201222995d64.png)

**Reply by jtkech** [reply]
Add `.ListAsync()` at the end ;)

**Reply by orchardc** [reply]
That has the same issue :(

` var contentItems = await _session.Query<ContentItem, ContentItemIndex>(x => x.ContentType == "MyProduct").ListAsync();`

**Reply by jtkech** [reply]
E.g. we have the following in the source code 

    var taxonomies = await _session.Query<ContentItem, ContentItemIndex>(q => q.ContentType == "Taxonomy" && q.Published).ListAsync();

**Reply by orchardc** [reply]
Thanks @jtkech , it works now. 👍  However, the table **IndexingTask** doesn't seem to be cleaned up (as I'm using Search function)

**Reply by jtkech** [reply]
The `IndexingTask` is not an index table mapped to an index provider as the other ones, look at the `IndexingTaskManager` that updates this table by using explicitly a new transaction

This table is used by the lucene worker feature (need to be enabled) through a background task this to update lucene indexes based on the content item ids in this table, but only for the new ones that was updated

To do this, the table `Id` column is used as a cursor, we have a `lucene.status.json` in which we save per lucene index the last `Id` that was updated, so that the lucene worker knows that it needs to take into account all records whose Id > last Id.

So if the lucene worker feature is not enabled, this table is only updated but never used. If the feature is enabled, normally the worker will not try to re-index a given content item id in a given record whose Id < last Id in the json file

Hmm, but if you remove `lucene.status.json`, it may try to restart from the beginning, didn't try it but i think it would not fail

Otherwise, to remove the related records you would need to do it explicitly on this table

**Reply by orchardc** [reply]
@jtkech, thanks for the information, very helpful.

What I found is: although the Delete and Create new record works (thanks for your guys help!!)

After new records created, the Lucene Search doesn't work - I think it's because indices not built yet.

Is it possible to programmatically **rebuild** the Lucene Indices? Just want to trigger it immediately after records creation.

![image](https://user-images.githubusercontent.com/16193516/103049597-1ed4d480-458a-11eb-912c-0324528ceb33.png)

**Reply by jtkech** [reply]
You can try to use the `LuceneIndexingService` scoped service, needs a reference on `OC.Lucene`
See what the `LuceneIndexingService` does in its `ResetIndex()` and `RebuildIndexAsync()` methods

**Reply by orchardc** [reply]
It works well, thank you @jtkech , my only concern now is the **IndexingTask** table will ever-growing, which might impact performance. 
I think it should be deleted as part of the _session.delete() process.

**Reply by jtkech** [reply]
Or a contentManager `HardDeleteAsync()` that would call handlers, one of the handler being implemented by the `OC.Indexing` module to remove the related item from the `IndexingTask` table.

**Comment by orchardc** [comment]
I think OC should proivde a way to remove (physically) content records from the DB.

Looking into tables: there might only be 5 tables which contains content:

> AliasPartIndex
> AutoroutePartIndex
> ContentItemIndex
> Document
> IndexingTask

**Reply by Skrypt** [reply]
This will be available after the Audit Trail module will be merged. The idea is to be able to see every changes that has been made on a content item. For elements "soft deleted" we could have a "recycle bin" which would allow to remove the items completely or restore them.

**Reply by ShaneCourtrille** [reply]
@Skrypt Is there a work item for permanent deletion now that Audit Trail is in?

**Reply by Skrypt** [reply]
None yet. It is up for grabs by anyone interested.
